File: ZxPlugins-lib/platform/win10/ZxPluginLib/src/BaseCommand.cs
using System;
using System.Reflection;
using System.Diagnostics;
using System.Collections.Generic;

namespace CocosAppWinRT.ZxPlugin
{
    /// <summary>
    /// Command base for create windows platform plugins
    /// </summary>
    /// <remarks>All platform plugins must extend BaseCommand</remarks>
    public class BaseCommand : IDisposable
    {
        /// <summary>
        /// Eventhandler for custom Command result
        /// </summary>
        public event EventHandler<CommandResult> OnCommandResult;

        /// <summary>
        /// Property for store current command callback id
        /// </summary>
        public string CurrentCommandCallbackId { get; set; }

        /// <summary>
        /// default constructor
        /// </summary>
        public BaseCommand()
        {
            ResultHandlers = new Dictionary<string, EventHandler<CommandResult>>();
        }

        /// <summary>
        /// Dictionary for store result handlers for commands with callback id
        /// </summary>
        protected Dictionary<string, EventHandler<CommandResult>> ResultHandlers;
        /// <summary>
        /// Method for add result handler
        /// </summary>
        /// <param name="callbackId">Callback id</param>
        /// <param name="handler">handler</param>
        public void AddResultHandler(string callbackId, EventHandler<CommandResult> handler)
        {
            ResultHandlers.Add(callbackId, handler);
        }
        /// <summary>
        /// Method for remove result handler
        /// </summary>
        /// <param name="callbackId">Callback id</param>
        /// <returns>True if can remove.</returns>
        public bool RemoveResultHandler(string callbackId)
        {
            return ResultHandlers.Remove(callbackId);
        }

        /// <summary>
        /// Method for invoke plugin command method.
        /// InvokeMethodNamed will call the named method of a BaseCommand subclass if it exists and pass the variable arguments list along.
        /// </summary>
        /// <param name="callbackId">Callback id</param>
        /// <param name="methodName">plugin method name to invoke</param>
        /// <param name="args">method arguments</param>
        /// <returns>Invoked oject result</returns>
        public object InvokeMethodNamed(string callbackId, string methodName, params object[] args)
        {
            //Debug.WriteLine(string.Format("InvokeMethodNamed:{0} callbackId:{1}",methodName,callbackId));
            this.CurrentCommandCallbackId = callbackId;
            return InvokeMethodNamed(methodName, args);
        }

        /// <summary>
        /// Method for invoke plugin command method.
        /// InvokeMethodNamed will call the named method of a BaseCommand subclass if it exists and pass the variable arguments list along.
        /// </summary>
        /// <param name="methodName">plugin method name to invoke</param>
        /// <param name="args">method arguments</param>
        /// <returns>Invoked oject result</returns>
        public object InvokeMethodNamed(string methodName, params object[] args)
        {
            MethodInfo mInfo = this.GetType().GetMethod(methodName);

            if (mInfo != null)
            {
                // every function handles DispatchCommandResult by itself
                return mInfo.Invoke(this, args);
            }

            // actually methodName could refer to a property
            if (args == null || args.Length == 0 ||
               (args.Length == 1 && "undefined".Equals(args[0])))
            {
                PropertyInfo pInfo = this.GetType().GetProperty(methodName);
                if (pInfo != null)
                {
                    object res = pInfo.GetValue(this, null);

                    DispatchCommandResult(new CommandResult(CommandResult.Status.OK, res));

                    return res;
                }
            }

            throw new MissingMethodException(methodName);

        }

        /// <summary>
        /// Method for dispatch command result
        /// </summary>
        public void DispatchCommandResult()
        {
            this.DispatchCommandResult(new CommandResult(CommandResult.Status.NO_RESULT));
        }

        /// <summary>
        /// Method for dispatch command result with values
        /// </summary>
        /// <param name="_result">command result</param>
        /// <param name="_callbackId">Callback id for command</param>
        public void DispatchCommandResult(CommandResult _result, string _callbackId="")
        {
            if (!string.IsNullOrEmpty(_callbackId)) 
            {
                _result.CallbackId = _callbackId;
            }
            else
            {
                _result.CallbackId = this.CurrentCommandCallbackId;
            }

            if (ResultHandlers.ContainsKey(_result.CallbackId))
            {
                ResultHandlers[_result.CallbackId](this, _result);
            }
            else if (this.OnCommandResult != null)
            {
                OnCommandResult(this, _result);
            }
            else
            {
                Debug.WriteLine("Failed to locate callback for id : " + _result.CallbackId);
            }

            if (!_result.KeepCallback)
            {
                this.Dispose();
            }

        }


        /// <summary>
        /// Occurs when the application is being deactivated.
        /// </summary>        
        public virtual void OnReset() {}

        /// <summary>
        /// Occurs when the application is being loaded, and the config.xml has an autoload entry
        /// </summary>    
        public virtual void OnInit() {}


        /// <summary>
        /// Occurs when the application is being made active after previously being put
        /// into a dormant state or tombstoned.
        /// </summary>        
        public virtual void OnActivated(object sender, Windows.UI.Core.WindowActivatedEventArgs e) { }

        /// <summary>
        /// Dispose command
        /// </summary>
        public void Dispose()
        {
            //Window.Current.Activated -= this.OnActivated;
            this.OnCommandResult = null;
        }

        /// <summary>
        /// Detach all command handlers
        /// </summary>
        public void DetachHandlers()
        {
            this.OnCommandResult = null;
            foreach (string callbackId in new List<string>(ResultHandlers.Keys))
            {
                RemoveResultHandler(callbackId);
            }
        }

    }

}

File: ZxPlugins-lib/platform/android/zxpluginlib/src/main/java/com/dgzornoza/zxplugin/BaseCommand.java
package com.dgzornoza.zxplugin;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 Command base for create windows platform plugins
 @remarks All platform plugins must extend BaseCommand
 */
public class BaseCommand
{
    private String m_currentCommandCallbackId;

    /**
    Dictionary for store result handlers for commands with callback id
    */
    protected Map<String, EventHandler<CommandResult>> ResultHandlers;

    /**
     Eventhandler for custom Command result
    */
    public EventHandler<CommandResult> OnCommandResult;

    public String getCurrentCommandCallbackId() { return this.m_currentCommandCallbackId; }
    public void setCurrentCommandCallbackId(String _callbackId) { this.m_currentCommandCallbackId = _callbackId; }

    /**
     default constructor
    */
    public BaseCommand()
    {
        ResultHandlers = new HashMap<String, EventHandler<CommandResult>>();
    }


    /**
    Method for add result handler
    @param _callbackId Callback id
    @param _handler handler
     */
    public void AddResultHandler(String _callbackId, EventHandler<CommandResult> _handler)
    {
        ResultHandlers.put(_callbackId, _handler);
    }
    /**
     Method for remove result handler
     @param _callbackId Callback id
     @return True if can remove.
     */
    public boolean RemoveResultHandler(String _callbackId)
    {
        boolean result = false;
        if(ResultHandlers.containsKey(_callbackId))
        {
            ResultHandlers.remove(_callbackId);
            result = true;
        }

        return result;
    }

    /**
    Method for invoke plugin command method.
    InvokeMethodNamed will call the named method of a BaseCommand subclass if it exists and pass the variable arguments list along.
    @param _callbackId Callback id
    @param _methodName plugin method name to invoke
    @param _args method arguments
    @return Invoked object result
    */
    public Object InvokeMethodNamed(String _callbackId, String _methodName, Object[] _args) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException
    {
        //Logger.getLogger(getClass().getName()).log(Level.INFO, "InvokeMethodNamed:" + _methodName + " callbackId: " + _callbackId);
        this.m_currentCommandCallbackId = _callbackId;
        return InvokeMethodNamed(_methodName, _args);
    }

    /**
     Method for invoke plugin command method.
     InvokeMethodNamed will call the named method of a BaseCommand subclass if it exists and pass the variable arguments list along.
     @param_ methodName plugin method name to invoke
     @param _args method arguments
     @return Invoked oject result
     */
    public Object InvokeMethodNamed(String _methodName, Object[] _args) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {

        Method mInfo = this.getClass().getMethod(_methodName);
        return mInfo.invoke(this, _args);
    }

    /**
     Method for dispatch command result
     */
    public void DispatchCommandResult()
    {
        this.DispatchCommandResult(new CommandResult(CommandResult.Status.NO_RESULT));
    }

    /**
     Method for dispatch command result with values
     @param _result command result
     */
    public void DispatchCommandResult(CommandResult _result)
    {
        this.DispatchCommandResult(_result, "");
    }

    /**
     Method for dispatch command result with values
     @param _result command result
     @param _callbackId Callback id for command
     */
    public void DispatchCommandResult(CommandResult _result, String _callbackId)
    {
        if (null != _callbackId && _callbackId.trim().length() > 0)
        {
            _result.setCallbackId(_callbackId);
        }
        else
        {
            _result.setCallbackId(this.m_currentCommandCallbackId);
        }

        if (ResultHandlers.containsKey(_result.getCallbackId()))
        {
            ResultHandlers.get(_result.getCallbackId()).Invoke(this, _result);
        }
        else if (this.OnCommandResult != null)
        {
            OnCommandResult.Invoke(this, _result);
        }
        else
        {
            Logger.getLogger(getClass().getName()).log(Level.WARNING, "Failed to locate callback for id : " + _result.getCallbackId());
        }

        if (!_result.getIsKeepCallback())
        {
            this.OnCommandResult = null;
        }

    }


    /**
     Occurs when the application is being deactivated.
    */
    public void OnReset() {}

    /**
     Occurs when the application is being loaded, and the config.xml has an autoload entry
    */
    public void OnInit() {}

    /**
     Detach all command handlers
    */
    public void DetachHandlers()
    {
        this.OnCommandResult = null;

        Set<String> keys = new HashSet<>(ResultHandlers.keySet());
        for (String callbackId : keys )
        {
            RemoveResultHandler(callbackId);
        }
    }

}
